clang/test/Analysis/fread.c

```diff
@@ -0,0 +1,412 @@
+#define PastMaxComplexity MaxInvalidatedElementRegion + 1
+void test_large_read(void) {
+  int buffer[PastMaxComplexity + 1];
+  buffer[PastMaxComplexity] = 42;
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    if (buffer[PastMaxComplexity] != 42) {
+      clang_analyzer_warnIfReached(); // Unreachable.
+    }
+    if (1 == fread(buffer, sizeof(int), PastMaxComplexity, fp)) {
+      if (buffer[PastMaxComplexity] != 42) {
+        clang_analyzer_warnIfReached(); // expected-warning{{REACHABLE}}
+      }
+    }
+    fclose(fp);
+  }
+}
+
+void test_small_read(void) {
+  int buffer[10];
+  buffer[5] = 42;
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    clang_analyzer_dump(buffer[5]); // expected-warning{{42 S32b}}
+    if (1 == fread(buffer, sizeof(int), 5, fp)) {
+      clang_analyzer_dump(buffer[5]); // expected-warning{{42 S32b}}
+    }
+    fclose(fp);
+  }
+}
+
+void test_partial_elements_read(void) {
+  clang_analyzer_dump(sizeof(int)); // expected-warning {{4 S32b}}
+
+  int buffer[100];
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    // 3*5: 15 bytes read; which is not exactly 4 integers, thus we invalidate the whole buffer.
+    if (5 == fread(buffer + 1, 3, 5, fp)) {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{derived_}}
+    } else {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{derived_}}
+    }
+    fclose(fp);
+  }
+}
+
+void test_whole_elements_read(void) {
+  clang_analyzer_dump(sizeof(int)); // expected-warning {{4 S32b}}
+
+  int buffer[100];
+  buffer[0] = 1;
+  buffer[15] = 2;
+  buffer[16] = 3;
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    // 3*20: 60 bytes read; which is basically 15 integers.
+    if (20 == fread(buffer + 1, 3, 20, fp)) {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{1 S32b}}
+      clang_analyzer_dump(buffer[20]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[21]); // expected-warning{{1st function call argument is an uninitialized value}}
+    } else {
+      clang_analyzer_dump(buffer[0]); // expected-warning{{1 S32b}}
+      clang_analyzer_dump(buffer[20]); // expected-warning{{conj_}}
+      clang_analyzer_dump(buffer[21]); // expected-warning{{1st function call argument is an uninitialized value}}
+    }
+    fclose(fp);
+  }
+}
+
+void test_unaligned_start_read(void) {
+  clang_analyzer_dump(sizeof(int)); // expected-warning {{4 S32b}}
+
+  int buffer[100];
+  buffer[0] = 3;
+  buffer[1] = 4;
+  buffer[2] = 5;
+  char *asChar = (char*)buffer;
+
+  FILE *fp = fopen("/home/test", "rb+");
+  if (fp) {
+    // We have an 'int' binding at offset 0 of value 3.
+    // We read 4 bytes at byte offset: 1,2,3,4.
+    if (4 == fread(asChar + 1, 1, 4, fp)) {
+      void clang_analyzer_printState(void);
+      clang_analyzer_printState();
+      clang_analyzer_dump(buffer[0]); // expected-warning{{3 S32b}} FIXME Reading a 'char' should not result in a 'S32b' value.
```

Comment:
The FIXME is not understandable to me, if this is about the problem seen in the PR remark (problem with store and overlapping read) it would be better to make this comment more accurate.

Reply:
I tried to highlight that reading a 8 bit wide `char` element should never result in a value of 32 bits.
And this is casued by the current Store implementation, as: it does a lookup and it returns whatever is associated with that read start bit offset. Given that the last write the Store saw was the 32 bit wide int 3, it will just return that - no matter if that bitwidth would make sense or not.

I've added the 8 bit assumption to the FIXME. I hope it's good now.